src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -791,9 +793,22 @@ private async Task EnsureInitializeAsync()
             try
             {
                 // If already initialized, need not be on the UI thread
-                if (!_initialized)
+                if (_initialized)
                 {
-                    _initialized = true;
+                    await EnsureNuGetAndVsProjectAdapterCacheAsync();
+                    return;
+                }
+
+                try
+                {
+                    // Ensure all initialization finished when needed, it still runs as async and prevents _initialized set true too early.
+                    // Setting '_initialized = true' too early caused random timing bug.
+                    await SemaphoreLock.WaitAsync();
```

Comment:
I think this can deadlock. 

Say the UI thread is blocked on WaitAsync, while another thread that acquired the lock wants to switch to the UI thread at line 813. 
The task that's actually doing the initialization needs to be part of the joinable task collection in order to mitigate this. 

SemaphoreSlim is not JTF aware, so I think you either need to join the task or use ReentrantSemaphore.

Reply:
@nkolev92
Can you refer to existing example for join the task approach?

Reply:
I don't know if we do it in our repo anywhere, you probably want to look at the APIs first, at microsoft/vs-threading or something along those lines. 
I know we use ReentrantSemaphore in the PM UI  (please search the codebase). 

idk how you know about JTF, but there's a one note page with some links to some talks explaining how it works. I've sent you the link offline.

Reply:
@nkolev92 
Can you check now? I implemented Reentrant lock.